### PR TITLE
reactive programming on java - rxjava

### PR DESCRIPTION
**Pull request by winstonxie3000**

### CODE REVIEW

File: _posts/2018-12-27-reactive-programming-in-java.markdown [hunk 1]
```diff
@@ -0,0 +1,110 @@
+# Reactive programming in Java
+
+Nowdays the notion of reactive programming is becoming more and more popular. So what is reactive programming, and what is the problem it is trying to resolve? Let's get started by the defintion in Wikipedia "Reactive Programming is an asynchronous programming paradigm concerned with data streams and the propagation of change."
```
Comment (u9520107) [inline _posts/2018-12-27-reactive-programming-in-java.markdown]: by the definition in Wikipedia ... => by "its" definitions in Wikipidia":" ...

File: _posts/2018-12-27-reactive-programming-in-java.markdown [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+### Why do we need the reactive programming?
```
Comment (u9520107) [inline _posts/2018-12-27-reactive-programming-in-java.markdown]: Why do we need reactive programming?
"the" is unnecessary here.

File: _posts/2018-12-27-reactive-programming-in-java.markdown [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+The simple answer is that we want to make our application more responsive to improve the user experience. There is one important word in the reactive programming definition: asynchronous. You are notified when data is emitted in the stream asynchronously – meaning independently to the main program flow. By structuring your program around data streams, you are writing asynchronous code: you write code invoked when the stream emits a new item. Threads, blocking code and side-effects are very important matters in this context.
```
Comment (u9520107) [inline _posts/2018-12-27-reactive-programming-in-java.markdown]: you write code "that is" invoked when...
Perhaps change the last sentence to "The concepts of threads, blocking code, and side-effects are very important in this context.

File: _posts/2018-12-27-reactive-programming-in-java.markdown [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+In the traditional programming model, there is one dilemma we need to face which is ***Blocking***. The modern applications usually need to accommodate a huge numbers of concurrent users, so the performance is becoming a key conern. In the traditional ways, we often program using blocking code. and when there is a performance bottleneck the normal way we do is introducing additional threads runing with similar blocking code which is very wastefull. Compared to the threads model, the reactive programming is non-blocking, asynchronous and resilient.
```
Comment (u9520107) [inline _posts/2018-12-27-reactive-programming-in-java.markdown]: there is one dilemma we need to face... => there is one dilemma we need to face":" Blocking.
a huge "number" of concurrent users
so performance is becoming a key concern. ("the" is not necessary here)
In traditional ways ("the" is also not necessary here)
"When" there is a performance bottleneck, normally we introduce additional threads "running" similar blocking code"," which is very "wasteful".  (The "and" at the beginning is not necessary)

File: _posts/2018-12-27-reactive-programming-in-java.markdown [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+### When do we use reactive programming and what is the benefit?
+The key expected benefit of reactive and non-blocking is the ability to scale with a small, fixed number of threads and less memory. That makes applications more resilient under load because they scale in a more predictable way. It allows you to treat streams of asynchronous events with the same sort of simple, composable operations that you use for collections of data items like arrays. It frees you from tangled webs of callbacks, and thereby makes your code more readable and less prone to bugs.
+
+### Reactive extension
+In software programming, Reactive Extensions is a set of tools allowing imperative programming languages to operate on sequences of data regardless of whether the data is synchronous or asynchronous. It provides a set of sequence operators that operate on each item in the sequence. Like ReactiveX, it's not only a API but provide the library for composing asynchronous and event-based programs by using observable sequences. it includes a set of lanuage implementation like RxJava, RxJS, RxCPP, etc.
```
Comment (u9520107) [inline _posts/2018-12-27-reactive-programming-in-java.markdown]: "It" includes a set ...

File: _posts/2018-12-27-reactive-programming-in-java.markdown [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+```
+<dependency>
+    <groupId>io.reactivex</groupId>
+    <artifactId>rxjava</artifactId>
+    <version>${rx.java.version}</version>
+</dependency>
+```
+There are two key types to understand when working with RxJava:
+
+- Observable represents any object that can get data from a data source and whose state may be of interest in a way that other objects may register an interest
```
Comment (u9520107) [inline _posts/2018-12-27-reactive-programming-in-java.markdown]: "An" observable ... whose state may be of interest "to other objects".

File: _posts/2018-12-27-reactive-programming-in-java.markdown [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+- An observer is any object that wishes to be notified when the state of another object changes
```
Comment (u9520107) [inline _posts/2018-12-27-reactive-programming-in-java.markdown]: object changes"."

File: _posts/2018-12-27-reactive-programming-in-java.markdown [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+20. t
+21. u
+22. v
+23. w
+24. x
+25. y
+26. z
+```
+### Conclusion
+We've talked about the concept of reactive programming, the benefit, how can we use it and different implementation and take RxJava as the example. Basically, the reactive programming is a concept and methodology, different people or group might have different view or interpretation on it but one simple word on the purpose is to make the application more responsive to provide the high performance to our users
```
Comment (u9520107) [inline _posts/2018-12-27-reactive-programming-in-java.markdown]: We've talked about the concept of reactive programming, the "benefits", how "we can" use it"," and different implementations "using RxJava" as the example.
Basically, reactive programming ... ("the" is not necessary)..., different people for "groups" might have different "views" or "interpretations" "of" it. But the simple purpose if reactive programming is to make the application more responsive to provide high performance applications to our users.

File: _posts/2018-12-27-reactive-programming-in-java.markdown [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+String[] letters = {"a", "b", "c", "d", "e", "f", "g"};
+Observable<String> observable = Observable.from(letters);
+observable.subscribe(
+  i -> result += i,  //OnNext
+  Throwable::printStackTrace, //OnError
+  () -> result += "_Completed" //OnCompleted
+);
+assertTrue(result.equals("abcdefg_Completed"));
+```
+The samle for itertor:
```
Comment (alvita) [inline _posts/2018-12-27-reactive-programming-in-java.markdown]: samle-> sample

File: _posts/2018-12-27-reactive-programming-in-java.markdown (hunk 8 above)
Comment (winstonxie3000) [inline _posts/2018-12-27-reactive-programming-in-java.markdown]: @u9520107 resolved all the grammar issues, please check again. Thanks a lot for your suggestion.

File: _posts/2018-12-27-reactive-programming-in-java.markdown [hunk 10]
```diff
@@ -1,17 +1,17 @@
 # Reactive programming in Java
 
-Nowdays the notion of reactive programming is becoming more and more popular. So what is reactive programming, and what is the problem it is trying to resolve? Let's get started by the defintion in Wikipedia "Reactive Programming is an asynchronous programming paradigm concerned with data streams and the propagation of change."
+Nowdays the notion of reactive programming is becoming more and more popular. So what is reactive programming, and what is the problem it is trying to resolve? Let's get started by its defintion in Wikipedia "Reactive Programming is an asynchronous programming paradigm concerned with data streams and the propagation of change."
```
Comment (u9520107) [inline _posts/2018-12-27-reactive-programming-in-java.markdown]: let's add a ':' after Wikipedia.

File: _posts/2018-12-27-reactive-programming-in-java.markdown [hunk 11]
```diff
@@ -107,4 +107,4 @@ with the result:
 26. z
 ```
 ### Conclusion
-We've talked about the concept of reactive programming, the benefit, how can we use it and different implementation and take RxJava as the example. Basically, the reactive programming is a concept and methodology, different people or group might have different view or interpretation on it but one simple word on the purpose is to make the application more responsive to provide the high performance to our users
+We've talked about the concept of reactive programming, the benefits, how we can use it, and different implementations using RxJava as the example. Basically, reactive programming is a concept and methodology, different people or groups might have different views or interpretations on it but one simple word on the purpose is to make the application more responsive to provide the high performance to our users
```
Comment (u9520107) [inline _posts/2018-12-27-reactive-programming-in-java.markdown]: let's put a ',' after "different people or groups might have different views or interpretations on it"

**Pull request merged**